test/test.ml

```diff
@@ -0,0 +1,95 @@
+open Hpack
+
+type entry = Header of header | Size of int
+
+let header = (module struct
+  type t = header
+
+  let pp fmt {name; value; never_index} =
+    Format.fprintf fmt "@[<hv 2>{ name = %S;@ value = %S;@ never_index = %s; }@]"
+      name value (if never_index then "true" else "false")
+
+  let equal = ( = )
+end : Alcotest.TESTABLE with type t = header)
+
+let encoding = (module struct
+  type t = string
+
+  let pp fmt = Format.fprintf fmt "%S"
+
+  let equal = ( = )
+end : Alcotest.TESTABLE with type t = string)
+
+let rec headers_of_entries = function
+  | Header header :: entries -> header :: headers_of_entries entries
+  | Size _ :: entries -> headers_of_entries entries
+  | [] -> []
+
+let test entries s = [
+  "Encode", `Quick, begin fun () ->
+    let t = Faraday.create 100 in
+    let encoder = Encoder.create () in
+    entries |> List.iter begin function
+    | Header header -> Encoder.encode_header encoder t header
+    | Size size -> Encoder.change_table_size encoder size
+    end;
+    let s' = Faraday.serialize_to_string t in
+    Alcotest.(check encoding) "same encoding" s s'
+  end;
+  "Decode", `Quick, begin fun () ->
+    let decoder = Decoder.create () in
+    Alcotest.(check (result (list header) reject))
+      "same headers"
+      (Angstrom.parse_string (Decoder.headers decoder) s)
+      (Ok (headers_of_entries entries))
+  end;
+]
+
+let test_eviction =
+  let entries = [
+    Size 38;
+    Header {name = "ABC"; value = "XYZ"; never_index = false};
+    Header {name = "ABCD"; value = "XYZ"; never_index = false};
+    Header {name = "ABC"; value = "XYZ"; never_index = false};
+    Header {name = "ABC"; value = "XYZ"; never_index = false};
+    Header {name = "ABCD"; value = "XYZ"; never_index = false};
+  ] in
+  let s =
+    "\x3f\x07\
+     \x40\x03ABC\x03XYZ\
+     \x40\x04ABCD\x03XYZ\
+     \x40\x03ABC\x03XYZ\
+     \xbe\
+     \x40\x04ABCD\x03XYZ" in
+  test entries s
+
+let test_size_update =
+  let entries = [
+    Header {name = "ABC"; value = "XYZ"; never_index = false};
+    Header {name = "ABC"; value = "XYZ"; never_index = false};
+    Size 0; Size 100;
```

Comment:
is this testing for the following?

> Multiple updates to the maximum table size can occur between the transmission of two header blocks. In the case that this size is changed more than once in this interval, the smallest maximum table size that occurs in that interval MUST be signaled in a dynamic table size update.

from [RFC7541§4.2](https://httpwg.org/specs/rfc7541.html#rfc.section.4.2)

it would be useful to add some references to the RFC for when other people (or future you 🙂) look at this in the future

Reply:
That's correct, I'll add references.

Reply:
thanks!